### PR TITLE
operationalize creating initial config file

### PR DESCRIPTION
**Pull request by gouthamr321**
This just takes chris's original script and operationalizes it and adds a few changes

notable changes:
- use an argument parser so things arent hard coded
- includes a helper function to sort round names
- Saves config file in json format instead of as a yaml format. This is because yaml files don't preserve the order of keys in a dictionary. See here: [https://github.com/yaml/pyyaml/issues/110](here). I'm not sure if saving this will cause issues downstream..... though should be easy to modify other scripts to read in the json file instead)

The output config files are here: /allen/aics/assay-dev/users/Goutham/4iProcessing-/4iprocessing/json_configs....... These look consistent to me compared to the original yaml files, but I'm hoping @cfrick13 can confirm

### CODE REVIEW

**Review by BrianWhitneyAI — APPROVED** [review]

File: 4iprocessing/generate_initial_config.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+parser.add_argument('--input_dirs', type=str, default=["/allen/aics/microscopy/Antoine/Analyse EMT/4i Data/5500000733","/allen/aics/microscopy/Antoine/Analyse EMT/4i Data/5500000724", "/allen/aics/microscopy/Antoine/Analyse EMT/4i Data/5500000728", "/allen/aics/microscopy/Antoine/Analyse EMT/4i Data/5500000726", "/allen/aics/microscopy/Antoine/Analyse EMT/4i Data/5500000725"], help="input dirs to parse")
+parser.add_argument('--output_path', type=str, required=True, help="final alignment outputs to specify in yaml file")
+parser.add_argument('--output_yaml_dir', type=str, required=True, help="final alignment outputs to specify in yaml file")
+
+
+def sort_rounds(rounds):
+    # function to sort rounds
+    Orginal_numbered_list = []
+    for i in range(len(rounds)):
+        int(re.search(r'\d+', rounds[i]).group())
```
Comment (cfrick13) [inline 4iprocessing/generate_initial_config.py]: this line is superfluous i think

File: 4iprocessing/generate_initial_config.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                    fpath2 = fpathr
+                    fpath = fpath2.replace("'",'')
+
+                    # each round of imaging is a multi-scene file like with dimensions STCZYX (only "time lapse rounds" have T>1). All rounds have S>1 and T>1
+                    # define the location of the reference channel in the list of channels
+                    # ideally the reference channel will be the image containing the nuclear fluorescence (or perhaps brightfield)
+                    # the reference image type (e.g. nuclear dye or brightfield) should be the consisently chosen for all rounds
+                    # if the image is a timelapse or round01 then choose the reference channel to be the last channel in the image set , =-1.
+
+                    ref_channel = -2
```
Comment (cfrick13) [inline 4iprocessing/generate_initial_config.py]: lines 74 - 79 should be moved below line 84

File: 4iprocessing/generate_initial_config.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                    if bool(re.search('time|1(?![0-9])',round_num, re.IGNORECASE)) and round_num!="Round 11":
```
Comment (cfrick13) [inline 4iprocessing/generate_initial_config.py:101]: need to make this catch any round greater than 10. right now it only looks for round 11 but in theory we may have 12 or 13 rounds in the future.

File: 4iprocessing/generate_initial_config.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                            scenes_to_toss.append(si+1)
+
+                    zscenes_to_toss=','.join([str(x) for x in scenes_to_toss])
+                    zchannels = ','.join(channels)
+                    #the names for these yaml entries are weird (i.e. "iround") because they get organized alphabetically and I want them to appear in a desired order
+                    print(f"channels are {zchannels}")
+                    detailid={}
+                    detailid['round'] = round_num
+                    detailid['item'] = 'czi'
+                    detailid['zchannels'] = zchannels
```
Comment (cfrick13) [inline 4iprocessing/generate_initial_config.py]: change name to `channels` instead of `zchannels`

Comment (cfrick13) [reply 4iprocessing/generate_initial_config.py]: I also want channels to be at the bottom, so maybe move this line below 107.

File: 4iprocessing/generate_initial_config.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                    detailid['path'] = fpath
+                    detailid['scenes_to_toss'] = zscenes_to_toss
+                    detailid['ref_channel'] = str(channels[ref_channel])
+                    config['Data'].append(detailid)
+                # yamd['Data'].append(subd)
+            config['barcode'] = barcode
+            config['scope'] = scope
+            
+            # output path defies folder where all images get stored after they get processed
```
Comment (cfrick13) [inline 4iprocessing/generate_initial_config.py]: typo: "defies"--> "defines"

**Review by cfrick13 — COMMENTED** [review]
made some comments in the code. looks good overall

File: 4iprocessing/generate_initial_config.py (hunk 1 above)
Comment (gouthamr321) [reply 4iprocessing/generate_initial_config.py]: Nice catch, line 24 is superfluous

File: 4iprocessing/generate_initial_config.py (hunk 2 above)
Comment (gouthamr321) [reply 4iprocessing/generate_initial_config.py]: This doesn't make any functional changes.... we are just defining the reference channel which doesn't affect the reading of the file..... but maybe I'm not understanding?

File: 4iprocessing/generate_initial_config.py (hunk 4 above)
Comment (gouthamr321) [reply 4iprocessing/generate_initial_config.py]: ohh, nice catch

File: 4iprocessing/generate_initial_config.py (hunk 2 above)
Comment (cfrick13) [reply 4iprocessing/generate_initial_config.py]: yes. no functional changes. it just makes sense to put the code where we choose channels after we have determined the channels from reader.channels. so its just a cosmetic change.

File: 4iprocessing/generate_initial_config.py (hunk 3 above)
Comment (gouthamr321) [reply 4iprocessing/generate_initial_config.py:101]: Just confirming, here we only what this if statement to be executed if we are looking at the first round or the 20X timelapse.... is that correct?

Comment (cfrick13) [reply 4iprocessing/generate_initial_config.py:101]: correct

File: 4iprocessing/generate_initial_config.py (hunk 5 above)
Comment (gouthamr321) [reply 4iprocessing/generate_initial_config.py]: Addressed in new commits

**Comment by cfrick13** [comment]
we talked and decided to wait to merge this code until we finish cleaning up variable names in the code (in `FormattingAndSorting` branch). And also decided to look at other ways of writing yamls to violate alphabetical order

File: 4iprocessing/generate_initial_config.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+import re
+from aicsimageio import AICSImage
+import argparse
+import json
+
+# run this to make the initial yaml files and then edit those files to make sure nothing is missing.
+# this code only helps make the yaml files...it does not generate a perfect yaml automatically. 
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--input_dirs', type=str, default=["/allen/aics/microscopy/Antoine/Analyse EMT/4i Data/5500000733","/allen/aics/microscopy/Antoine/Analyse EMT/4i Data/5500000724", "/allen/aics/microscopy/Antoine/Analyse EMT/4i Data/5500000728", "/allen/aics/microscopy/Antoine/Analyse EMT/4i Data/5500000726", "/allen/aics/microscopy/Antoine/Analyse EMT/4i Data/5500000725"], help="input dirs to parse")
```
Comment (aditnath) [inline 4iprocessing/generate_initial_config.py]: might want to not have these hardcoded as these will change once the isilon is gone?

File: 4iprocessing/generate_initial_config.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        config['Data']=[]
+        scope_list = [x for x in os.listdir(bdir) if 'ZSD' in x]
+        for scope in scope_list: # for each scope
+            pdir = bdir + os.sep + scope
+            round_list = [x for x in os.listdir(pdir) if bool(re.search('Time|Round [0-9]+',x,re.IGNORECASE))&(Path(x).stem==Path(x).name)]
+            round_list = sort_rounds(round_list)
+            print("round list is {}".format(round_list))
+            for round_num in round_list:
+                ppath = os.path.join(pdir, round_num)
+                czi_list = [x for x in os.listdir(ppath) if ('.czi' in x)&bool(re.search('20x',x,re.IGNORECASE))]
```
Comment (aditnath) [inline 4iprocessing/generate_initial_config.py]: we will need to include support for ome.tiffs as well, since those are the aligned files

**Review by aditnath — CHANGES_REQUESTED** [review]
left a couple comments, we might want to address the ome tiff functionality before merging

File: 4iprocessing/generate_initial_config.py (hunk 6 above)
Comment (gouthamr321) [reply 4iprocessing/generate_initial_config.py]: nice catch, took care of this

File: 4iprocessing/generate_initial_config.py (hunk 7 above)
Comment (gouthamr321) [reply 4iprocessing/generate_initial_config.py]: also just took care of this one

**Comment by gouthamr321** [comment]
Incorporated some changes since last time. Notable changes are:
- workaround to save config file in yaml format without messing up order of elements in dictionary
- defined some operations as separate functions(e.g. determining which scenes to toss, determining reference channel)
- support for files saved in .tiff format for the future
- change to argument parser to not hardcode input directory files 

Not sure if this is ready to merge or if we still need to wait for FormattingAndSorting branch to be merged first? Thoughts?

**Comment by cfrick13** [comment]
We should be ready to merge this now. FormattingAndSorting branch is merged

**Review by BrianWhitneyAI — APPROVED** [review]

**Review by cfrick13 — APPROVED** [review]
looks good. Might just need change requests marked as completed

**Review by aditnath — APPROVED** [review]

**Comment by gouthamr321** [comment]
Operationalize config file generation

**Pull request merged**